test/connection-timeout.js

```diff
@@ -73,6 +73,28 @@ describe('connection timeout', () => {
     })
   })
 
+  it('should not break further pending checkouts on a timeout', (done) => {
+    const pool = new Pool({ connectionTimeoutMillis: 200, max: 1 })
+    pool.connect((err, client, releaseOuter) => {
+      expect(err).to.be(undefined)
+
+      pool.connect((err, client) => {
+        expect(err).to.be.an(Error)
+        expect(client).to.be(undefined)
+        releaseOuter()
+      })
+
+      setTimeout(() => {
+        pool.connect((err, client, releaseInner) => {
+          expect(err).to.be(undefined)
+          expect(client).to.not.be(undefined)
+          releaseInner()
+          pool.end(done)
+        })
+      }, 100)
```

Comment:
I think this can fail with a race condition. What about:

```js
pool.connect((err, client, releaseOuter) => {
  expect(err).to.be(undefined)

  pool.connect((err, client) => {
    expect(err).to.be.an(Error)
    expect(client).to.be(undefined)
    releaseOuter()

    setTimeout(() => {
      pool.connect((err, client, releaseInner) => {
        expect(err).to.be(undefined)
        expect(client).to.not.be(undefined)
        releaseInner()
        pool.end(done)
      })
    }, 100)
  })

})
```

Reply:
The point of the test is to have two pending `connect`s at the same time. There’s a race condition in the sense that Node doesn’t make any guarantees about the ordering of the two timers, but given that they’re started simultaneously with timeouts of 100 and 200 ms, I think it’s fine for a test.